lib/src/component_declaration/function_component.dart

```diff
@@ -0,0 +1,181 @@
+// Copyright 2020 Workiva Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// ignore_for_file: prefer_generic_function_type_aliases
+library over_react.component_declaration.function_component;
+
+import 'package:js/js_util.dart';
+import 'package:meta/meta.dart';
+import 'package:over_react/over_react.dart';
+import 'package:react/react.dart' as react;
+import 'package:react/react_client.dart';
+import 'package:react/react_client/js_backed_map.dart';
+
+export 'component_type_checking.dart'
+    show isComponentOfType, isValidElementOfType;
+
+/// A React component declared using a function that takes in [props] and returns rendered output.
+///
+/// See <https://facebook.github.io/react/docs/components-and-props.html#functional-and-class-components>.
+typedef dynamic UiFunctionComponent<T extends UiProps>(T props);
+
+/// Returns a `UiFactory<TProps>` for [functionComponent].
+///
+/// __Example__:
+/// ```dart
+/// UiFactory<FooProps> Foo = uiFunction((props) {
+///     // Set default props using null-aware operators.
+///     final isDisabled = props.isDisabled ?? false;
+///     final items = props.items ?? [];
+///
+///     // Return the rendered component contents here.
+///     return Dom.div()(items);
+///   },
+///   // The generated props config will match the factory name.
+///   $FooPropsConfig,
+/// );
+///
+/// mixin FooProps on UiProps {
+///   bool isDisabled;
+///   Iterable<String> items;
+/// }
+/// ```
+///
+/// __OR__ Optionally pass in an existing [PropsFactory] in place of a props [config].
+///
+/// ```dart
+/// UiFactory<FooProps> Bar = uiFunction((props) {
+///     return Dom.div()(props.items);
+///   },
+///   null,
+///   propsFactory: PropsFactory.fromUiFactory(Foo),
+///   displayName: 'Bar',
+/// );
+/// ```
+///
+/// __OR__ Set [config] to `null` when using `UiProps`.
+///
+/// ```dart
+/// UiFactory<UiProps> Baz = uiFunction((props) {
+///     return Dom.div()('prop id: ${props.id}');
+///   },
+///   null,
+///   displayName: 'Baz',
+/// );
+/// ```
+///
+/// Learn more: <https://reactjs.org/docs/components-and-props.html#function-and-class-components>.
+// FIXME right now only top level factory declarations will generate props configs.
+UiFactory<TProps> uiFunction<TProps extends UiProps>(
+  UiFunctionComponent<TProps> functionComponent,
+  FunctionComponentConfig<TProps> config, {
```

Comment:
Since these are generated, it feels like `FunctionComponentConfig` should either be private, or `$$` prefixed or something to indicate its not intended for public consumption _(and subsequently not exported)_

Reply:
I'm not sure I agree. I think this is similar to `PropsMeta`; most of the time it's generated, but there's no reason to prevent people from consuming it.

Reply:
How would someone create their own?

Reply:
Not sure I follow; what would prevent them from creating their own?

Reply:
How / why would you create an instance of a config like this when it is something that our builder generates?

Reply:
How: just instantiate it 
Why: If you want to use a `uiFunction` that doesn't have a generated builder

It's definitely an edge-case that I can't think of any concrete use-cases for, but I also see no reason to prohibit it.

On a related note: we have this config, and then two arguments that map to the same fields as the config. Perhaps we should just combine those?

Existing usages could be replaced like so:

```diff
-uiFunction(() ..., null, displayName: 'Foo');
+uiFunction(() ..., GenericUiPropsConfig(displayName: 'Foo'));
```
```diff
-uiFunction(
-  () ..., 
-  null, 
-  propsFactory: PropsFactory.fromUiFactory(Bar),
-  displayName: 'Foo',
);
+uiFunction(
+  () ..., 
+  FunctionComponentConfig(
+    propsFactory: PropsFactory.fromUiFactory(Bar),
+    displayName: 'Foo',
+  ),
+);
```

Then, we would have a legitimate use for people doing this.

Thoughts, @sydneyjodon-wk  / @aaronlademann-wf ?

Reply:
Not a fan of requiring them to use the `FunctionComponentConfig` to set a `displayName`.

Reply:
I agree with Aaron and the `UiProps` case seems a little over complicated.

But, now that I think about it, we could generate a config for every function component so that the display name will be set automatically and just not set `propsFactory` in the config if `propsFactory` is already set on `uiFunction` or if `UiProps` is used because those are taken care of inside `uiFunction`. Thoughts?

Then we could change the name to `$$FunctionComponentConfig` or something like Aaron was saying before since the config would always be generated.

Reply:
> Not a fan of requiring them to use the FunctionComponentConfig to set a displayName.
That's fair, but keep in mind this is a bit of an edge case already.

> I agree with Aaron and the UiProps case seems a little over complicated.
That's fair; my thought was that it'd a little more straightforward than passing in `null`.

> we could generate a config for every function component 
Don't we already do that though for generated function components?

This case is for non-generated function components, many of which may be local variables, and I don't think we should generate for those.

Reply:
So should I update this to have propsFactory and displayName set in a `FunctionComponentConfig` or just leave `FunctionComponentConfig` as an optional way to set them?

Reply:
I would lean toward FunctionComponentConfig and get rid of the args, but I'd be curious to hear other people's opinions. Maybe we should ask in Slack?

Reply:
I'll ask in slack

Reply:
Thank you!

Reply:
I updated this!